File: 253.meeting-rooms-ii.java
/*
 * @lc app=leetcode id=253 lang=java
 *
 * [253] Meeting Rooms II
 */
import java.util.*;

// @lc code=start
class Solution {
    public int minMeetingRooms(int[][] intervals) {
        final int n = intervals.length;
        int[] start = new int[n], end = new int[n];
        for (int i = 0; i < n; ++i) {
            start[i] = intervals[i][0];
            end[i] = intervals[i][1];
        }

        Arrays.sort(start);
        Arrays.sort(end);
        int rooms = 0, endId = 0;
        for (int i = 0; i < n; ++i) {
            if (start[i] < end[endId]) {
                ++rooms;    // overlap - need a new room
            } else {
                // no overlap - the current interval can reuse an existing room
                ++endId;
            }
        }
        return rooms;
    }
}
// @lc code=end

File: 253.meeting-rooms-ii.go
/*
 * @lc app=leetcode id=253 lang=golang
 *
 * [253] Meeting Rooms II
 */
import "sort"

// @lc code=start
// Sort interval starts and ends separately, and then check overlaps
// T: O(n * log(n))
// S: O(n) for copying
func minMeetingRooms(intervals [][]int) int {
	n := len(intervals)

	starts := make([]int, n)
	ends := make([]int, n)
	for i := range intervals {
		starts[i] = intervals[i][0]
		ends[i] = intervals[i][1]
	}

	sort.Ints(starts)
	sort.Ints(ends)

	curEndIdx := 0
	ans := 0
	for i := 0; i < n; i++ {
		if starts[i] < ends[curEndIdx] {
			// overlap, need a new room
			ans++
		} else {
			// no overlap, reuse an existing room
			curEndIdx++
		}
	}
	return ans
}

// @lc code=end

